File: node/parser.js
const { TokenType } = require('./lexer');
const { NumberNode, BinaryOpNode, UnaryOpNode, RenderNode } = require('./ast');

class Parser {
  constructor(tokens) {
    this.tokens = tokens;
    this.position = 0;
    this.currentToken = this.tokens[this.position];
  }

  advance() {
    this.position++;
    if (this.position < this.tokens.length) {
      this.currentToken = this.tokens[this.position];
    }
  }

  factor() {
    const token = this.currentToken;

    if (token.type === TokenType.PLUS || token.type === TokenType.MINUS || token.type === TokenType.FACTORIAL) {
      this.advance();
      return new UnaryOpNode(token.type, this.factor());
    }

    if (token.type === TokenType.NUMBER) {
      this.advance();
      return new NumberNode(token.value);
    }

    if (token.type === TokenType.LPAREN) {
      this.advance();
      const node = this.expression();
      if (this.currentToken.type !== TokenType.RPAREN) {
        throw new Error('Expected closing parenthesis');
      }
      this.advance();
      return node;
    }

    throw new Error(`Unexpected token: ${token.type}`);
  }

  power() {
    let node = this.factor();

    while (this.currentToken.type === TokenType.EXPONENT) {
      const token = this.currentToken;
      this.advance();
      node = new BinaryOpNode(node, token.type, this.factor());
    }

    return node;
  }

  term() {
    let node = this.power();

    while (this.currentToken.type === TokenType.MULTIPLY || 
           this.currentToken.type === TokenType.DIVIDE ||
           this.currentToken.type === TokenType.MODULO) {
      const token = this.currentToken;
      this.advance();
      node = new BinaryOpNode(node, token.type, this.power());
    }

    return node;
  }

  expression() {
    let node = this.term();

    while (this.currentToken.type === TokenType.PLUS || 
           this.currentToken.type === TokenType.MINUS) {
      const token = this.currentToken;
      this.advance();
      node = new BinaryOpNode(node, token.type, this.term());
    }

    return node;
  }

  statement() {
    if (this.currentToken.type === TokenType.RENDER) {
      this.advance();
      if (this.currentToken.type !== TokenType.LPAREN) {
        throw new Error('Expected opening parenthesis after render');
      }
      this.advance();
      const expr = this.expression();
      if (this.currentToken.type !== TokenType.RPAREN) {
        throw new Error('Expected closing parenthesis');
      }
      this.advance();
      return new RenderNode(expr);
    }
    return this.expression();
  }

  parse() {
    const ast = this.statement();
    if (this.currentToken.type !== TokenType.EOF) {
      throw new Error(`Unexpected token at end: ${this.currentToken.type}`);
    }
    return ast;
  }
}

module.exports = { Parser };

File: ruby/parser.rb
require_relative 'ast'

# Parser for calculator language using recursive descent
class Parser
  def initialize(tokens)
    @tokens = tokens
    @position = 0
  end

  def parse
    # Check if this is a render expression
    if current_token && current_token.type == :RENDER
      consume(:RENDER)
      expr = expression
      result = AST::RenderNode.new(expr)
    else
      result = expression
    end
    
    if @position < @tokens.length
      raise "Unexpected token after expression: #{current_token.value}"
    end
    result
  end

  private

  def expression
    assignment_expression
  end

  def assignment_expression
    expr = additive_expression

    # Check if this is an assignment (identifier = expression)
    if current_token && current_token.type == :ASSIGN && expr.is_a?(AST::VariableNode)
      consume(:ASSIGN)
      value_expr = assignment_expression
      return AST::AssignmentNode.new(expr.name, value_expr)
    end

    expr
  end

  def additive_expression
    left = multiplicative_expression

    while current_token && [:PLUS, :MINUS].include?(current_token.type)
      operator = current_token.value
      consume(current_token.type)
      right = multiplicative_expression
      left = AST::BinaryOpNode.new(left, operator, right)
    end

    left
  end

  def multiplicative_expression
    left = unary_expression
    while current_token && [:MULTIPLY, :DIVIDE, :MODULO, :EXPONENT].include?(current_token.type)
      operator = current_token.value
      consume(current_token.type)
      right = unary_expression
      left = AST::BinaryOpNode.new(left, operator, right)
    end

    left
  end

  def unary_expression
    if current_token && current_token.type == :MINUS
      consume(:MINUS)
      operand = unary_expression
      AST::UnaryOpNode.new('-', operand)
    elsif current_token && current_token.type == :FACTORIAL
      consume(:FACTORIAL)
      operand = unary_expression
      AST::UnaryOpNode.new('!', operand)
    else
      primary_expression
    end
  end

  def primary_expression
    if current_token && current_token.type == :INTEGER
      value = current_token.value
      consume(:INTEGER)
      AST::IntegerNode.new(value)
    elsif current_token && current_token.type == :IDENTIFIER
      name = current_token.value
      consume(:IDENTIFIER)
      AST::VariableNode.new(name)
    elsif current_token && current_token.type == :LPAREN
      consume(:LPAREN)
      expr = expression
      consume(:RPAREN)
      expr
    else
      raise "Expected integer, identifier, or '(' but got #{current_token ? current_token.value : 'EOF'}"
    end
  end

  def current_token
    @position < @tokens.length ? @tokens[@position] : nil
  end

  def consume(expected_type)
    if current_token.nil?
      raise "Expected #{expected_type} but reached end of input"
    elsif current_token.type != expected_type
      raise "Expected #{expected_type} but got #{current_token.type}"
    end
    @position += 1
  end
end